src/freenet/support/io/Fallocate.java

```diff
@@ -38,49 +38,69 @@ private Fallocate(FileChannel channel, int fd, long final_filesize) {
   }
 
   public static Fallocate forChannel(FileChannel channel, long final_filesize) {
-    return new Fallocate(channel, getDescriptor(channel), final_filesize);
+    try {
+      return new Fallocate(channel, getDescriptor(channel), final_filesize);
+    } catch (final UnsupportedOperationException exception) {
+      // File descriptor is not supported: fd is null and unavailable
+      return new Fallocate(channel, 0, final_filesize);
+    }
   }
 
   public static Fallocate forChannel(FileChannel channel, FileDescriptor fd, long final_filesize) {
-    return new Fallocate(channel, getDescriptor(fd), final_filesize);
+    try {
+      return new Fallocate(channel, getDescriptor(fd), final_filesize);
+    } catch (final UnsupportedOperationException exception) {
+      // File descriptor is not supported: fd is null and unavailable
+      return new Fallocate(channel, 0, final_filesize);
+    }
   }
 
-  public Fallocate fromOffset(long offset) {
+  public Fallocate fromOffset(long offset) throws IllegalArgumentException {
     if(offset < 0 || offset > final_filesize) throw new IllegalArgumentException();
     this.offset = offset;
     return this;
   }
 
-  public Fallocate keepSize() {
+  public Fallocate keepSize() throws UnsupportedOperationException {
     requireLinux("fallocate keep size");
     mode |= FALLOC_FL_KEEP_SIZE;
     return this;
   }
 
-  private void requireLinux(String feature) {
+  private void requireLinux(String feature) throws UnsupportedOperationException {
     if (!IS_LINUX) {
       throwUnsupported(feature);
     }
   }
 
-  private void throwUnsupported(String feature) {
+  private void throwUnsupported(String feature) throws UnsupportedOperationException {
     throw new UnsupportedOperationException(feature + " is not supported on this file system");
   }
 
   public void execute() throws IOException {
     int errno = 0;
     boolean isUnsupported = false;
-    if (IS_LINUX) {
-      final int result = FallocateHolder.fallocate(fd, mode, offset, final_filesize-offset);
-      errno = result == 0 ? 0 : Native.getLastError();
-    } else if (IS_POSIX) {
-      errno = FallocateHolderPOSIX.posix_fallocate(fd, offset, final_filesize-offset);
+    if (fd != 0) {
+      if (IS_LINUX) {
+        final int result = FallocateHolder.fallocate(fd, mode, offset, final_filesize-offset);
+        errno = result == 0 ? 0 : Native.getLastError();
+      } else if (IS_POSIX) {
+        errno = FallocateHolderPOSIX.posix_fallocate(fd, offset, final_filesize-offset);
+      } else {
+        isUnsupported = true;
+      }
     } else {
-      isUnsupported = true;
+      // fd is null and unavailable
+      if (Platform.isWindows()) {
+        // Windows do not create sparse files by default, so just write a byte at the end.
```

Comment:
The point of `fallocate` is to efficiently pre-allocate disk space for sparse files (allocate the blocks, but don't write the zeroes).

If Windows does not create sparse files anyways, then why do we need this special case? We already deal with the unsupported case through `legacyFill` below.

Reply:
`legacyFill` fills the file with random data. It's wasting SSD life time because when the useful data get in, the random data have to be erased. However it does not need to erase when writing zeros. So I suggest `legacyFill` should fill the file with zeros. And for Windows, this is better, because there's only one API call and does not use a large buffer, so it's faster (suggested by https://stackoverflow.com/questions/18031841/pre-allocating-drive-space-for-file-storage referenced in this file).

Reply:
Depending on filesystem (and most notably full-disk encryption) the SSD erase argument would indeed be a valid one.

As far as I remember, the "fill with random data" approach was to prevent transparent disk compression from getting in the way of allocating sufficient space. But it may have become less relevant over time with ever growing disk capacity.

If we keep this special case for Windows, can you it to where the unsupported case is handled? (probably where `legacyFill()` is called) Then it is clear that this is not an equivalent of `fallocate`, but rather an optimized alternative for when `fallocate` is unsupported.